paasta_tools/autoscaling_lib.py

```diff
@@ -514,41 +547,165 @@ def spotfleet_scaler(resource, error):
+                    if is_ready_to_kill:
+                        log.info("TERMINATING: {0}".format(instance_id))
+                        try:
+                            ec2_client.terminate_instances(InstanceIds=[instance_id], DryRun=dry_run)
+                        except ClientError as e:
+                            if e.response['Error'].get('Code') == 'DryRunOperation':
+                                pass
+                            else:
+                                raise
+                        instances_to_kill.pop(slave_pid)
+                    else:
+                        log.debug("Instance {0}: NOT ready to kill".format(instance_id))
+                log.debug("Waiting 5 seconds and then checking again")
+                time.sleep(5)
+    except TimeoutError:
+        log.error("Timed out after {0} waiting to drain {1}, now terminating anyway".format(CLUSTER_DRAIN_TIMEOUT,
+                                                                                            instances_to_kill))
+        try:
+            ec2_client.terminate_instances(InstanceIds=instances_to_kill, DryRun=dry_run)
+        except ClientError as e:
+            if e.response['Error'].get('Code') == 'DryRunOperation':
+                pass
+            else:
+                raise
+
+
+@register_autoscaling_component('aws_spot_fleet_request', SCALER_KEY)
+def scale_aws_spot_fleet_request(resource, delta, target_capacity, sorted_slaves, dry_run):
+    """Scales a spot fleet request by delta to reach target capacity
+    If scaling up we just set target capacity and let AWS take care of the rest
+    If scaling down we pick the slaves we'd prefer to kill, put them in maintenance
+    mode and drain them (via paasta_maintenance and setup_marathon_jobs). We then kill
+    them once they are running 0 tasks or once a timeout is reached
+
+    :param resource: resource to scale
+    :param delta: integer change in number of servers
+    :param target_capacity: target number of instances
+    :param sorted_slaves: list of slaves by order to kill
+    :param dry_run: Don't drain or make changes to spot fleet if True"""
+    sfr_id = resource['id']
+    ec2_client = boto3.client('ec2')
+    if delta == 0:
+        return
+    elif delta > 0:
+        log.debug("Increasing spot fleet capacity to: {0}".format(target_capacity))
+        if not dry_run:
+            ec2_client.modify_spot_fleet_request(SpotFleetRequestId=sfr_id, TargetCapacity=target_capacity,
+                                                 ExcessCapacityTerminationPolicy='noTermination')
+            return
+    elif delta < 0:
+        number_to_kill = delta * -1
+        sfr_ips = get_sfr_instance_ips(sfr_id)
+        log.debug("IPs in SFR: {0}".format(sfr_ips))
+        sfr_sorted_slaves = [slave for slave in sorted_slaves if slave_pid_to_ip(slave['pid']) in sfr_ips]
+        log.debug("SFR slave kill preference: {0}".format(sfr_sorted_slaves))
```

Comment:
can you just print the slave ids here, rather than the entire object? this log line is going to be very hard to parse otherwise